File: system_monitor/collector/remote.py
from __future__ import annotations

import json
import threading
import time
from typing import Any

import httpx

from .base import Sensor, SensorReading


def _extract_json_path(data: Any, path: str) -> Any:
    current = data
    for part in path.split("."):
        if isinstance(current, dict):
            current = current.get(part)
        else:
            return None
    return current


class HttpJsonSensor(Sensor):
    def read(self) -> SensorReading:
        url = self.config.params.get("url")
        json_path = self.config.params.get("json_path", "")
        timeout = float(self.config.params.get("timeout", 5))

        if not url:
            return SensorReading(
                sensor_id=self.id,
                value=None,
                status="error",
                error="Не указан URL в params.url",
            )

        try:
            response = httpx.get(url, timeout=timeout)
            response.raise_for_status()
            payload = response.json()
            raw = _extract_json_path(payload, json_path) if json_path else payload
            value = float(raw)
        except Exception as exc:
            return SensorReading(
                sensor_id=self.id,
                value=None,
                status="error",
                error=str(exc),
            )

        return SensorReading(
            sensor_id=self.id,
            value=round(value, 3),
            status=self.evaluate_status(value),
        )


class _MqttListener:
    _instance: _MqttListener | None = None
    _lock = threading.Lock()

    def __init__(self) -> None:
        self._values: dict[str, float] = {}
        self._topics: dict[str, tuple[str, int]] = {}
        self._client: Any = None
        self._thread: threading.Thread | None = None
        self._started = False

    @classmethod
    def instance(cls) -> _MqttListener:
        with cls._lock:
            if cls._instance is None:
                cls._instance = cls()
            return cls._instance

    def register(self, cache_key: str, topic: str, broker: str, port: int) -> None:
        self._topics[cache_key] = (topic, broker, port)
        self._ensure_running()

    def get(self, cache_key: str) -> float | None:
        return self._values.get(cache_key)

    def _ensure_running(self) -> None:
        if self._started:
            return
        try:
            import paho.mqtt.client as mqtt  # type: ignore[import-untyped]
        except ImportError:
            return

        def on_message(_client: Any, _userdata: Any, msg: Any) -> None:
            topic = msg.topic
            for key, (registered_topic, _, _) in self._topics.items():
                if registered_topic != topic:
                    continue
                try:
                    payload = msg.payload.decode("utf-8")
                    try:
                        data = json.loads(payload)
                        if isinstance(data, dict) and "value" in data:
                            self._values[key] = float(data["value"])
                        else:
                            self._values[key] = float(data)
                    except (json.JSONDecodeError, TypeError, ValueError):
                        self._values[key] = float(payload)
                except (UnicodeDecodeError, ValueError):
                    pass

        client = mqtt.Client()
        client.on_message = on_message
        first = next(iter(self._topics.values()), None)
        if first is None:
            return
        broker, port = first[1], first[2]
        client.connect(broker, port, keepalive=30)
        for topic, _, _ in self._topics.values():
            client.subscribe(topic)

        def run() -> None:
            client.loop_forever()

        self._client = client
        self._thread = threading.Thread(target=run, name="system-monitor-mqtt", daemon=True)
        self._thread.start()
        self._started = True
        time.sleep(0.1)


class MqttSensor(Sensor):
    def read(self) -> SensorReading:
        topic = self.config.params.get("topic")
        broker = self.config.params.get("broker", "localhost")
        port = int(self.config.params.get("port", 1883))

        if not topic:
            return SensorReading(
                sensor_id=self.id,
                value=None,
                status="error",
                error="Не указан topic в params.topic",
            )

        listener = _MqttListener.instance()
        listener.register(self.id, topic, broker, port)

        if listener._client is None:
            return SensorReading(
                sensor_id=self.id,
                value=None,
                status="error",
                error="Установите paho-mqtt: pip install paho-mqtt",
            )

        value = listener.get(self.id)
        if value is None:
            return SensorReading(
                sensor_id=self.id,
                value=None,
                status="unknown",
                error="Нет данных по MQTT-топику",
            )

        return SensorReading(
            sensor_id=self.id,
            value=round(value, 3),
            status=self.evaluate_status(value),
        )

File: internal/collector/remote.go
package collector

import (
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"
	"strings"
	"sync"
	"time"

	mqtt "github.com/eclipse/paho.mqtt.golang"
)

type httpJsonSensor struct{ baseSensor }

func (s httpJsonSensor) Read() SensorReading {
	url, _ := s.cfg.Params["url"].(string)
	jsonPath, _ := s.cfg.Params["json_path"].(string)
	timeout := 5.0
	if t, ok := s.cfg.Params["timeout"].(float64); ok {
		timeout = t
	}
	if url == "" {
		return SensorReading{SensorID: s.ID(), Status: "error", Error: "Не указан URL в params.url"}
	}
	client := &http.Client{Timeout: time.Duration(timeout * float64(time.Second))}
	resp, err := client.Get(url)
	if err != nil {
		return SensorReading{SensorID: s.ID(), Status: "error", Error: err.Error()}
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return SensorReading{SensorID: s.ID(), Status: "error", Error: err.Error()}
	}
	var payload interface{}
	if err := json.Unmarshal(body, &payload); err != nil {
		return SensorReading{SensorID: s.ID(), Status: "error", Error: err.Error()}
	}
	raw := payload
	if jsonPath != "" {
		raw = extractJSONPath(payload, jsonPath)
	}
	value, err := toFloatValue(raw)
	if err != nil {
		return SensorReading{SensorID: s.ID(), Status: "error", Error: err.Error()}
	}
	v := math.Round(value*1000) / 1000
	return SensorReading{SensorID: s.ID(), Value: &v, Status: evaluateStatus(s.cfg, &v)}
}

func extractJSONPath(data interface{}, path string) interface{} {
	current := data
	for _, part := range strings.Split(path, ".") {
		if m, ok := current.(map[string]interface{}); ok {
			current = m[part]
		} else {
			return nil
		}
	}
	return current
}

func toFloatValue(v interface{}) (float64, error) {
	switch n := v.(type) {
	case float64:
		return n, nil
	case int:
		return float64(n), nil
	case json.Number:
		return n.Float64()
	default:
		return 0, fmt.Errorf("cannot convert value to float")
	}
}

type mqttSensor struct{ baseSensor }

var mqttListener = &mqttListenerState{
	values: map[string]float64{},
	topics: map[string]mqttTopic{},
}

type mqttListenerState struct {
	mu      sync.Mutex
	values  map[string]float64
	topics  map[string]mqttTopic
	client  mqtt.Client
	started bool
}

type mqttTopic struct {
	topic  string
	broker string
	port   int
}

func (l *mqttListenerState) register(key, topic, broker string, port int) {
	l.mu.Lock()
	defer l.mu.Unlock()
	l.topics[key] = mqttTopic{topic: topic, broker: broker, port: port}
	l.ensureRunning()
}

func (l *mqttListenerState) get(key string) *float64 {
	l.mu.Lock()
	defer l.mu.Unlock()
	if v, ok := l.values[key]; ok {
		return &v
	}
	return nil
}

func (l *mqttListenerState) ensureRunning() {
	if l.started || len(l.topics) == 0 {
		return
	}
	if l.values == nil {
		l.values = map[string]float64{}
	}
	opts := mqtt.NewClientOptions()
	var first mqttTopic
	for _, t := range l.topics {
		first = t
		break
	}
	opts.AddBroker(fmt.Sprintf("tcp://%s:%d", first.broker, first.port))
	opts.SetDefaultPublishHandler(func(_ mqtt.Client, msg mqtt.Message) {
		l.mu.Lock()
		defer l.mu.Unlock()
		for key, t := range l.topics {
			if t.topic != msg.Topic() {
				continue
			}
			payload := string(msg.Payload())
			var data interface{}
			if err := json.Unmarshal(msg.Payload(), &data); err == nil {
				if m, ok := data.(map[string]interface{}); ok {
					if v, ok := m["value"]; ok {
						if f, err := toFloatValue(v); err == nil {
							l.values[key] = f
							continue
						}
					}
				}
				if f, err := toFloatValue(data); err == nil {
					l.values[key] = f
					continue
				}
			}
			if f, err := strconvParse(payload); err == nil {
				l.values[key] = f
			}
		}
	})
	l.client = mqtt.NewClient(opts)
	if token := l.client.Connect(); token.Wait() && token.Error() != nil {
		return
	}
	seen := map[string]bool{}
	for _, t := range l.topics {
		if seen[t.topic] {
			continue
		}
		seen[t.topic] = true
		l.client.Subscribe(t.topic, 0, nil)
	}
	l.started = true
}

func strconvParse(s string) (float64, error) {
	var v float64
	_, err := fmt.Sscanf(strings.TrimSpace(s), "%f", &v)
	return v, err
}

func (s mqttSensor) Read() SensorReading {
	topic, _ := s.cfg.Params["topic"].(string)
	broker, _ := s.cfg.Params["broker"].(string)
	if broker == "" {
		broker = "localhost"
	}
	port := 1883
	if p, ok := s.cfg.Params["port"].(int); ok {
		port = p
	} else if p, ok := s.cfg.Params["port"].(float64); ok {
		port = int(p)
	}
	if topic == "" {
		return SensorReading{SensorID: s.ID(), Status: "error", Error: "Не указан topic в params.topic"}
	}
	mqttListener.register(s.ID(), topic, broker, port)
	value := mqttListener.get(s.ID())
	if value == nil {
		return SensorReading{SensorID: s.ID(), Status: "unknown", Error: "Нет данных по MQTT-топику"}
	}
	v := math.Round(*value*1000) / 1000
	return SensorReading{SensorID: s.ID(), Value: &v, Status: evaluateStatus(s.cfg, &v)}
}
